### PR TITLE
Some improvements and V2 support.

### PR DESCRIPTION
**Pull request by stephpy**
- Add cache implementation.
- Can sort providers.
- Usage of ChainProvider.

Before merging this branch, we have to create a 1.0 branch of master since this PR is compatible only with swap ~2.0

### CODE REVIEW

**Comment by florianv** [comment]
Nice work, thanks!

> Before merging this branch, we have to create a 1.0 branch of master since this PR is compatible only with swap ~2.0

There is a 1.x branch.

File: DependencyInjection/FlorianvSwapExtension.php [hunk 1]
```diff
@@ -35,37 +36,58 @@ public function load(array $config, ContainerBuilder $container)
...
-                        break;
+        if (isset($config['cache'])) {
+            $this->loadCache($config['cache'], $container);
+        }
+    }
+
+    private function loadProviders(array $config, ContainerBuilder $container)
+    {
+        foreach ($config as $providerName => $providerConfig) {
+            switch ($providerName) {
```
Comment (florianv) [inline DependencyInjection/FlorianvSwapExtension.php]: The [NationalBankOfRomania](https://github.com/florianv/swap/blob/master/src/Provider/NationalBankOfRomaniaProvider.php) provider is missing

File: Resources/doc/index.md [hunk 2]
```diff
@@ -124,6 +117,50 @@ In order to add your custom providers implementing `Swap\ProviderInterface`, you
...
+            priority: 0
+```
+
+```xml
+<service id="acme_demo.provider.custom" class="Acme\DemoBundle\Provider\Custom">
+    <tag name="florianv_swap.provider" priority="5" />
+</service>
+```
+
+In this case, the order of providers will be:
```
Comment (florianv) [inline Resources/doc/index.md]: The order should be (from highest to lowest priority):

- yahoo_finance
- acme_demo.provider.custom
- google_finance

File: DependencyInjection/Configuration.php [hunk 3]
```diff
@@ -55,4 +70,49 @@ public function getConfigTreeBuilder()
...
+        $node = $treeBuilder->root($name);
+
+        $node
+            ->children()
+                ->integerNode('priority')->defaultValue(0)->end()
+            ->end()
+        ;
+
+        return $node;
+
```
Comment (florianv) [inline DependencyInjection/Configuration.php]: There is an extra line

**Comment by florianv** [comment]
It would be nice if the http adapter could be configurable:

```yaml
# app/config/config.yml
florianv_swap:
    http_adapter: service_id
```

with a default value corresponding to the `FileGetContentsHttpAdapter`

File: Resources/doc/index.md (hunk 2 above)
Comment (stephpy) [inline Resources/doc/index.md]: You're right ... It's how it work at this moment. :)

File: DependencyInjection/FlorianvSwapExtension.php (hunk 1 above)
Comment (stephpy) [inline DependencyInjection/FlorianvSwapExtension.php]: Ok, i copied previous bundle providers, i'll add it.

**Comment by stephpy** [comment]
:+1: for the http_adapter service.
I'll try to install today HHVM to fix tests (weird fail) and i'll fix these points.

**Comment by florianv** [comment]
Ok, thanks.

Otherwise everything looks good to me :+1:

**Comment by stephpy** [comment]
This is a known issue in HHVM:

https://github.com/facebook/hhvm/issues/2655

I saw than PHP 7 could do the same ... i'm currently downloading PHP 7 to try ...

IMO, By this way we have 2 solutions:

  - Using our system to SORT providers by priority without using usort ... (If PHP7 has the same behavior, we should consider it)
  - Allow failures on HHVM on travis

**Comment by stephpy** [comment]
@florianv it's OK, I used a simple system to sort providers. :)

File: DependencyInjection/FlorianvSwapExtension.php (hunk 1 above)
Comment (florianv) [inline DependencyInjection/FlorianvSwapExtension.php]: Thanks

**Comment by florianv** [comment]
> @florianv it's OK, I used a simple system to sort providers. :)

I think too it is better than dropping HHVM support.

File: DependencyInjection/Compiler/ProviderPass.php [hunk 4]
```diff
@@ -27,14 +28,48 @@ class ProviderPass implements CompilerPassInterface
      */
     public function process(ContainerBuilder $container)
     {
-        if (!$container->hasDefinition('florianv_swap.swap')) {
-            return;
+        $providers = array();
+
+        foreach ($container->findTaggedServiceIds('florianv_swap.provider') as $id => $attributes) {
+            if (count($attributes) > 1) {
```
Comment (florianv) [inline DependencyInjection/Compiler/ProviderPass.php]: Can't we just ignore the extra attributes ?

File: DependencyInjection/Compiler/ProviderPass.php [hunk 5]
```diff
@@ -27,14 +28,48 @@ class ProviderPass implements CompilerPassInterface
...
+                throw new \LogicException(sprintf('You have many `florianv_swap.provider` tag with id [%s]', $id));
+            }
+
+            $providers[$id] = isset($attributes[0]) ? $attributes[0] : 0;
+        }
+
+        $providers = $this->normalizeProviders($providers);
+
+        if (count($providers) === 1) {
```
Comment (florianv) [inline DependencyInjection/Compiler/ProviderPass.php]: I prefer to use `1 === count($providers)` as Symfony does

File: DependencyInjection/Compiler/ProviderPass.php [hunk 6]
```diff
@@ -27,14 +28,48 @@ class ProviderPass implements CompilerPassInterface
...
+    }
 
-        foreach ($container->findTaggedServiceIds('florianv_swap.provider') as $id => $attributes) {
-            $definition->addMethodCall('addProvider', array(new Reference($id)));
+    private function normalizeProviders(array $providers)
+    {
+        $providersByPriority = array();
+
+        foreach ($providers as $id => $provider) {
+            $priority                         = isset($provider['priority']) ? $provider['priority'] : 0;
```
Comment (florianv) [inline DependencyInjection/Compiler/ProviderPass.php]: Could you remove the alignments everywhere please (as it's done in Symfony)

File: Resources/config/services.xml [hunk 7]
```diff
@@ -6,18 +6,33 @@
...
+        <parameter key="florianv_swap.provider.chain.class">Swap\Provider\ChainProvider</parameter>
+        <parameter key="florianv_swap.provider.yahoo_finance.class">Swap\Provider\YahooFinanceProvider</parameter>
+        <parameter key="florianv_swap.provider.google_finance.class">Swap\Provider\GoogleFinanceProvider</parameter>
+        <parameter key="florianv_swap.provider.european_central_bank.class">Swap\Provider\EuropeanCentralBankProvider</parameter>
+        <parameter key="florianv_swap.provider.national_bank_of_romania.class">Swap\Provider\NationalBankOfRomaniaProvider</parameter>
+        <parameter key="florianv_swap.provider.open_exchange_rates.class">Swap\Provider\OpenExchangeRatesProvider</parameter>
+        <parameter key="florianv_swap.provider.xignite.class">Swap\Provider\XigniteProvider</parameter>
+        <parameter key="florianv_swap.provider.webservicex.class">Swap\Provider\WebserviceXProvider</parameter>
+
+
```
Comment (florianv) [inline Resources/config/services.xml]: There is an extra line

File: DependencyInjection/Configuration.php [hunk 8]
```diff
@@ -30,24 +30,41 @@ public function getConfigTreeBuilder()
...
             ->children()
+                ->scalarNode('http_adapter')->defaultValue('florianv_swap.http_adapter.file_get_contents')->end()
+                ->arrayNode('cache')
+                    ->children()
+                        ->integerNode('ttl')->isRequired()->end()
+                        ->append($this->getCacheDriverNode('doctrine'))
+                    ->end()
+                ->end()
                 ->arrayNode('providers')
+
```
Comment (florianv) [inline DependencyInjection/Configuration.php]: I think this extra line can be removed

**Comment by florianv** [comment]
Thanks @stephpy.

Looks good to me once the small CS issues are fixed.

If you could squash your commits too it would be perfect!

**Comment by stephpy** [comment]
Done ;)

**Comment by florianv** [comment]
Thanks

**Comment by stephpy** [comment]
Thanks ;)

**Pull request merged**